In [2]:
import numpy as np

class Dense_Layer:
    def __init__(self, inputs, weights, bias):
        self.inputs = np.array(inputs)
        self.weights = np.array(weights)
        self.bias = np.array(bias)
        self.output = None

    def forward(self):
        inp = np.array(self.inputs)
        weight = np.array(self.weights)
        bias = np.array(self.bias)

        if weight.ndim != 2:
            raise ValueError("weights must be a 2D array")

        n_in = inp.shape[-1]  # number of input features
        # Case A: weights shaped (n_in, n_out) -> use inp . W
        if weight.shape[0] == n_in:
            out = np.dot(inp, weight)
        # Case B: weights shaped (n_out, n_in) -> use W . inp
        elif weight.shape[1] == n_in:
            out = np.dot(weight, inp)
        else:
            raise ValueError(f"Weight shape {weight.shape} not compatible with input length {n_in}")

        # add bias (broadcast if needed)
        out = out + bias
        self.output = out
        return out

    def activate(self, func="relu"):
        if self.output is None:
            raise ValueError("Run forward() before activate()")
        if func.lower() == "relu":
            return np.maximum(0, self.output)
        elif func.lower() == "sigmoid":
            return 1 / (1 + np.exp(-self.output))
        elif func.lower() == "softmax":
            exp_vals = np.exp(self.output - np.max(self.output))
            return exp_vals / np.sum(exp_vals)
        else:
            raise ValueError("Unknown activation function")

    def loss(self, y_true, y_pred):
        y_true = np.array(y_true)
        y_pred = np.array(y_pred)
        return np.mean((y_true - y_pred)**2)

    

In [3]:
# Initial Input
X = [5.1, 3.5, 1.4, 0.2]
# Target Output
Target_Output = [0.7, 0.2, 0.1]

# First Hidden Layer
W1 = [[0.2, 0.1, -0.4, 0.6],
      [0.5, -0.2, 0.3, -0.1],
      [-0.3, 0.4, 0.2, 0.5]]   # shape (3x4)

B1 = [3.0, -2.1, 0.6]

# Second Hidden Layer

W2 = [[0.3, 0.7, -0.6],   
      [-0.5, 0.2, 0.4]]   # shape (2x3)
B2 = [4.3, 6.4]     


# Output Layer (3 neurons = 3 classes)
W3 = [[0.5, -0.2],
      [-0.3, 0.6], 
      [0.8, -0.4]]   # shape (3x2)
B3 = [-1.5, 2.1, -3.3]


In [4]:
layer1 = Dense_Layer(X, W1, B1)
a1 = layer1.activate("relu") if (layer1.forward() is not None) else None
print("Layer 1 output:", a1)

# Layer 2
layer2 = Dense_Layer(a1, W2, B2)
a2 = layer2.activate("sigmoid") if (layer2.forward() is not None) else None
print("Layer 2 output:", np.round(a2, 3))

# Output Layer
layer3 = Dense_Layer(a2, W3, B3)
a3 = layer3.activate("softmax") if (layer3.forward() is not None) else None
print("Final Output (probabilities):", np.round(a3, 3))

# Loss
loss_val = layer3.loss(Target_Output, a3)
print("Loss:", np.round(loss_val, 3))

Layer 1 output: [3.93 0.15 0.85]
Layer 2 output: [0.994 0.992]
Final Output (probabilities): [0.027 0.969 0.005]
Loss: 0.351


----------------------------------------------------------NUMBER 2----------------------------------------------------------

In [12]:
X2 = [14.1, 20.3, 0.095]
Target_Output2 = [1]

we1 = [[0.5, 0.2, -0.7],
       [-0.3, 0.4, 0.9],   #shape (3x3)
       [0.8, -0.6, 0.1]]

b1 = [0.3, -0.5, 0.6]

we2 = np.array([[0.6, -0.3],
                [-0.2, 0.5],
                [0.4, 0.7]])   # shape (3,2)

b2 = [4.3, 6.4]

we3 = np.array([[0.7],
                [-0.5]])   # shape (2,1)


In [15]:
layer11 = Dense_Layer(X2, we1, b1)
a11 = layer11.activate("relu") if (layer11.forward() is not None) else None
print("Layer 1 output:", a11)

# Layer 2
layer22 = Dense_Layer(a11, we2, b2)
a22 = layer22.activate("sigmoid") if (layer22.forward() is not None) else None
print("Layer 2 output:", np.round(a2, 4) )

# Output Layer
layer33 = Dense_Layer(a22, we3, b3)
a33 = layer3.activate("sigmoid") if (layer33.forward() is not None) else None
print("Final Output (probabilities):", np.round(a33, 3))

# Loss
loss_val1 = layer3.loss(Target_Output2, a33)
print("Loss:", np.round(loss_val1, 3))

Layer 1 output: [ 1.336  10.383   9.0095]
Layer 2 output: [0.9938 0.9919]
Final Output (probabilities): [0.231 0.917 0.052]
Loss: 0.499


Yes — the network predicts Malignant (1).
The final probability vector is [0.231, 0.917, 0.052] and the largest probability is 0.917 at index 1 - choose class 1 (Malignant).